ResearchKit/ActiveTasks/ORKTowerOfHanoiViewController.m

```diff
@@ -0,0 +1,253 @@
+    }
+    [self reloadData];
+    [self evaluatePuzzle];
+}
+
+#pragma Mark -- ORKTowerOfHanoiViewController
+
+- (void)updateTitleText {
+    NSString *text = [NSString stringWithFormat: ORKLocalizedString(@"TOWER_OF_HANOI_TASK_ACTIVE_STEP_PROGRESS_TEXT", nil), (int)_numberOfMoves, [self timeElapsedString]];
+    [self.activeStepView updateTitle:nil text: text];
+}
+
+- (NSString *)timeElapsedString {
+    NSDateComponentsFormatter *formatter = [NSDateComponentsFormatter new];
+    formatter.unitsStyle = NSDateComponentsFormatterUnitsStylePositional;
+    formatter.zeroFormattingBehavior = NSDateComponentsFormatterZeroFormattingBehaviorPad;
+    formatter.allowedUnits =  NSCalendarUnitMinute | NSCalendarUnitSecond;
+    return [formatter stringFromTimeInterval: _secondsElapsed];
+}
+
+- (void)reloadData {
+    for (ORKTowerOfHanoiTowerView *towerView in _towerViews) {
+        towerView.highlighted = _selectedIndex != nil && [_towerViews indexOfObject:towerView] == _selectedIndex.integerValue;
+        [towerView reloadData];
+    }
+}
+
+- (BOOL)puzzleIsSolved {
+    return ((ORKTowerOfHanoiTower *)_towers.lastObject).disks.count == _numberOfDisks;
+}
+
+- (void)evaluatePuzzle {
+    if ([self puzzleIsSolved]) {
+        [self finish];
+    }
+}
+
+- (void)setupTowers {
+    NSMutableArray *diskStack = [NSMutableArray array];
+    for (NSInteger disk = _numberOfDisks ; disk > 0 ; disk--) {
+        [diskStack addObject: @(disk)];
+    }
+    _towers = [@[[[ORKTowerOfHanoiTower alloc]initWithDisks:diskStack], [ORKTowerOfHanoiTower empty], [ORKTowerOfHanoiTower empty]] mutableCopy];
+}
+
+- (void)setupTowerViews {
+    NSMutableArray *towerViews = [NSMutableArray array];
+    for (NSInteger idx = 0 ; idx < 3 ; idx++) {
+        ORKTowerOfHanoiTowerView *towerView = [[ORKTowerOfHanoiTowerView alloc] initWithFrame:CGRectZero maximumNumberOfDisks:_numberOfDisks];
+        towerView.delegate = self;
+        towerView.dataSource = self;
+        towerView.targetTower = idx == 2;
+        [towerViews addObject:towerView];
+        [towerView setTranslatesAutoresizingMaskIntoConstraints:NO];
+        [_towerOfHanoiCustomView addSubview:towerView];
+    }
+    _towerViews = towerViews;
+}
+
+- (void)transferDiskFromTowerAtIndex:(NSInteger)donorTowerIndex toTowerAtIndex:(NSInteger)recipientTowerIndex {
+    ORKTowerOfHanoiTower *donorTower = _towers[donorTowerIndex];
+    ORKTowerOfHanoiTower *recipientTower = _towers[recipientTowerIndex];
+    if (donorTower.disks.count > 0 && [recipientTower canRecieveDisk:(NSNumber *)donorTower.disks.lastObject]) {
+        NSMutableArray *recipientDisks = [recipientTower.disks mutableCopy];
+        NSMutableArray *donorDisks = [donorTower.disks mutableCopy];
+        [recipientDisks addObject:donorDisks.lastObject];
+        [donorDisks removeLastObject];
+        _towers[recipientTowerIndex] = [[ORKTowerOfHanoiTower alloc]initWithDisks:recipientDisks];
```

Comment:
Is rebuilding the `ORKTowerOfHanoiTower` necessary?

I image implement this would save you a lot of hassle.
```
BOOL result = [recipientTower receiveDiskFrom: donorTower];
```